### PR TITLE
license should be from relative dir

### PR DESCRIPTION
**Pull request by gilescope**
As mentioned in #3352, if the license file is set and we're not in the root dir when calling `cargo fmt` it currently falls in a heap.

### CODE REVIEW

File: src/config/license.rs [hunk 1]
```diff
@@ -211,7 +211,23 @@ impl TemplateParser {
...
+    path: &str,
+) -> Result<Regex, LicenseError> {
+    let mut path = std::path::PathBuf::from(path);
+    if path.is_relative() {
+        if let Some(root_dir) = root_dir {
+            // The tempate path needs to be resolved relative to the .rustfmt config file
+            // rather than whatever dir cargo fmt was run from.
+            let current = std::env::current_dir().expect("pwd");
+            std::env::set_current_dir(root_dir).expect("alter pwd");
+            if let Ok(abs_path) = path.canonicalize() {
```
Comment (gilescope) [inline src/config/license.rs:225]: Hmm, https://github.com/rust-lang/rustfmt/pull/4131 suggests I use this instead:
```suggestion
            if let Ok(abs_path) = absolute_path(path) {
```

Comment (gilescope) [reply src/config/license.rs]: Hmm but then I see dunce https://github.com/rust-lang/rustfmt/pull/4137 and then at some point it is removed... - trying to figure out why.

Comment (gilescope) [reply src/config/license.rs:225]: shrugs. will go back to using canonicalize as that's used elsewhere in rustfmt. (Maybe canonicalize got better under windows?)

**Comment by gilescope** [comment]
@calebcartwright I can attest to this feature being pretty broken at the moment so this should be an easy review - it can't make the feature any worse.

**Comment by gilescope** [comment]
Can I help in any way or do anything more?

**Comment by calebcartwright** [comment]
> Can I help in any way or do anything more?

Not really. I appreciate the PR and that folks would probably prefer I get through issue and PR triage a bit more quickly :smile: 

However, the amount of work greatly overshadows the amount of bandwidth so the pace is what circumstances permit. I realize that this seems like a small/easy change, but path resolution changes have been a notorious source of bugs, often fixing one issue while creating another (especially on different platforms). It's not an area we have great tests for which ups the need for review/manual testing, and to be fully transparent, some of what I'll call rustfmt's auxiliary features (license checks, todo scans, etc.) are simply lower on the priority list relative to more core formatting issues.

> Hmm but then I see dunce #4137 and then at some point it is removed... - trying to figure out why.

Probably best summarized in https://github.com/rust-lang/rustfmt/pull/3914#issuecomment-917690028, but that's part of the story in the review. There's been lots of various changes around path resolution (both for licenses/configs as well as loading out-of-line mods during the earlier parsing phases) made on the other branch that I'll need to cross reference.

Sorry I don't have any better news, but promise I'll get to it when I can!

**Comment by calebcartwright** [comment]
Thanks @karyon, though none of those were what I was thinking of for this. I thought we had some other PRs that dealt with addressing config file resolutions in cases where there were multiple config files throughout the project. The dunce one is worth pursuing, and the others relate to some other areas that need attention, but I don't think of any of those will be applicable here.

If we are to go forward with this fix/enhancement to the license feature, I'd want to see some thorough test cases that cover the behavior. That should include scenarios with multiple config files with respective templates, as well as file-based definitions with command line overrides.

That being said, this is another config option, like `report_todo` and `report_fixme`, that just feels out of place for rustfmt. I understand the utility of the license check feature (and the other two as well), but I keep coming back to the conclusion those don't belong in rustfmt/would be better suited as a standalone tool.

**Comment by gilescope** [comment]
Just to be clear the licensing feature is pretty broken at the moment.

**Comment by calebcartwright** [comment]
> Just to be clear the licensing feature is pretty broken at the moment.

Thanks @gilescope, though to be honest I'm really not sure what to make of this given [your prior comment to this same effect](https://github.com/rust-lang/rustfmt/pull/4967#issuecomment-917448056) and the subsequent conversation.

To put it a little more pointedly, I'm seriously considering moving to drop this feature from rustfmt altogether, along with some of the other non-formatting related features that were stuffed into rustfmt for some reason.

However, if this is a feature we're going to keep, and if we're going to move forward with your proposed changes, then it would need to have some accompanying test cases to cover the various scenarios I described in https://github.com/rust-lang/rustfmt/pull/4967#issuecomment-955887425. I'm not requesting those tests at the moment though, as I wouldn't want you or anyone else to work on them in futility given the possible dropping/deprecation of this unstable feature.

**Comment by gilescope** [comment]
Agreed, do or do not :-)

Idk maybe it should live in cargo deny conceptually?

**Comment by calebcartwright** [comment]
Thanks again for the PR and bearing with us while we contemplated the future of the option. I've decided we're going to move forward with the deprecation path so I'm going to close this.

Refs #5103

**Pull request closed without merging**